### PR TITLE
Book controller

### PR DESCRIPTION
**Pull request by Parashchuk**

### CODE REVIEW

File: src/main/java/bookstore/config/MapperConfig.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import org.mapstruct.NullValueCheckStrategy;
+import org.mapstruct.ReportingPolicy;
+
+@org.mapstruct.MapperConfig(
+        componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        unmappedTargetPolicy = ReportingPolicy.IGNORE,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/config/MapperConfig.java]: it is better to specify @MapperConfig directly above the mapper class

File: src/main/java/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/api/books")
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

File: src/main/java/bookstore/dto/BookDto.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package bookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class BookDto {
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/dto/BookDto.java]: Let’s  use record for dtos

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+@RequiredArgsConstructor
+public class BookRepositoryImpl implements BookRepository {
+    private final SessionFactory factory;
+
+    @Override
+    public Book save(Book book) {
+        try {
+            factory.inTransaction(e -> e.persist(book));
+        } catch (Exception e) {
+            throw new RuntimeException("Save operation was unsuccessful: ", e);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: Let’s  use your custom DataProcessingException

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        return book;
+    }
+
+    public Optional<Book> findBookById(Long id) {
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java:28]: Override annotation is missed

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        try {
+            return Optional.ofNullable(factory.fromSession(e -> e.find(Book.class, id)));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: the same

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try {
+            return factory.fromSession(e -> e.createQuery("from Book", Book.class)
+                    .getResultList());
+        } catch (Exception e) {
+            throw new RuntimeException("Save operation was unsuccessful: ", e);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]: The same and be attentive with messages

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]:
```suggestion
```

File: src/main/java/bookstore/service/BookServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import bookstore.dto.CreateBookRequestDto;
+import bookstore.entity.Book;
+import bookstore.mapper.BookMapper;
+import bookstore.repository.BookRepositoryImpl;
+import jakarta.persistence.EntityNotFoundException;
+import java.util.List;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookServiceImpl.java]:
```suggestion
@Service
```
It’s  service

File: src/main/java/bookstore/service/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        return bookMapper.toDto(bookRepository.save(book));
+    }
+
+    @Override
+    public BookDto findBookById(Long id) {
+        Book book = bookRepository.findBookById(id).orElseThrow(
+                () -> new EntityNotFoundException("Book was not found with id: " + id)
+        );
+
+        return bookMapper.toDto(book);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookServiceImpl.java]: You can simplify  it using method map of class Optional

File: src/main/resources/application.properties [hunk 11]
```diff
@@ -1 +1,10 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/book-store
+spring.datasource.username=root
+spring.datasource.password=Root1234!
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.MySQLDialect
 
+spring.jpa.hibernate.ddl-auto=create-drop
+spring.jpa.show-sql=true
+
+spring.jackson.deserialization.fail-on-unknown-properties=true
```
Comment (liliia-ponomarenko) [inline src/main/resources/application.properties:10]: Why do you need this property?

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes ;)

File: src/main/resources/application.properties (hunk 11 above)
Comment (Parashchuk) [reply src/main/resources/application.properties:10]: when the req body comes with unknown params, then Jackson will return a 400 code response

File: src/main/java/bookstore/repository/BookRepositoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        return book;
+    }
+
+    @Override
+    public Optional<Book> findBookById(Long id) {
+        try {
+            return Optional.ofNullable(factory.fromSession(e -> e.find(Book.class, id)));
+        } catch (Exception e) {
+            throw new DataProcessingException("Search operation was unsuccessful: ", e);
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/repository/BookRepositoryImpl.java]:
```suggestion
            throw new EntityNotFoundException("Can’t  find book by id: " + id, e);
```

File: src/main/java/bookstore/service/BookServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+@Service
+@RequiredArgsConstructor
+public class BookServiceImpl implements BookService {
+    private final BookRepositoryImpl bookRepository;
+    private final BookMapper bookMapper;
+
+    @Override
+    public BookDto save(CreateBookRequestDto bookRequestDto) {
+        Book book = bookMapper.toModel(bookRequestDto);
+        return bookMapper.toDto(bookRepository.save(book));
```
Comment (liliia-ponomarenko) [inline src/main/java/bookstore/service/BookServiceImpl.java]: can be inline variable

File: src/main/resources/application.properties (hunk 11 above)
Comment (liliia-ponomarenko) [reply src/main/resources/application.properties:10]: which status will be returned without this property? I guess the same

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But fix a few mistakes before merging ;)

File: src/main/resources/application.properties (hunk 11 above)
Comment (Parashchuk) [reply src/main/resources/application.properties:10]: it'll just proceed if other data is valid. I don't know much about it, it was just recommended on the mate course

**Pull request merged**